### PR TITLE
Set file readonly if class is deployed or if server-side source control reports it not editable

### LINKED ISSUE #1397: Allow server-side file to readonly if server-side source control class for namespace returns that status
Some source control systems require code to be reserved/checked-out before it is changed.

Since ISFS was originally implemented VS Code's `FileSystemProvider` API has gained the ability for its `stat` method's [FileStat](https://code.visualstudio.com/api/references/vscode-api#FileStat) return value to include an optional `permissions` property of type [FilePermission](https://code.visualstudio.com/api/references/vscode-api#FilePermission). One of its bitmask values is `Readonly`.

I propose adding a boolean setting to make `stat` return `permissions` with the `Readonly` bit set or cleared depending on the value of the `Editable` boolean returned from the [GetStatus](https://docs.intersystems.com/irislatest/csp/documatic/%25CSP.Documatic.cls?LIBRARY=%25SYS&CLASSNAME=%25Studio.Extension.Base#GetStatus) method of the server-side source control class (if any) of the namespace.

The new setting would default to `false` so the new behaviour is opt-in and thus won't break existing environments.

It would be a resource-level setting so it can be configured server-side per namespace if desired.

### PR DESCRIPTION
**Pull request by gjsjohnmurray**
This PR closes #1397

In some circumstances the current VS Code (1.91.1) doesn't correctly switch an editor to be readonly. My PR https://github.com/microsoft/vscode/pull/221023 will resolve this. Meanwhile I am creating this PR as a Draft.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1312,6 +1312,11 @@
           "type": "boolean",
           "scope": "resource"
         },
+        "objectscript.serverSourceControl.respectEditableStatus": {
```
Comment (isc-bsaviano) [inline package.json:1315]: Can we add an explicit default of `false`?

File: package.json [hunk 2]
```diff
@@ -1312,6 +1312,11 @@
...
+          "description": "Set ISFS document readonly if GetStatus method of server-side source control class returns Editable=0",
```
Comment (isc-bsaviano) [inline package.json:1316]:
```suggestion
          "markdownDescription": "Set `isfs` document readonly if GetStatus method of server-side source control class returns Editable = 0.",
```

File: src/providers/FileSystemProvider/FileSystemProvider.ts [hunk 3]
```diff
@@ -184,13 +192,55 @@ export class FileSystemProvider implements vscode.FileSystemProvider {
...
+      result = await entryPromise;
+    } else {
+      // otherwise return the promise
+      return entryPromise;
+    }
+
+    //
+    if (result instanceof File) {
+      const api = new AtelierAPI(uri);
+      const serverName = getServerName(uri);
```
Comment (isc-bsaviano) [inline src/providers/FileSystemProvider/FileSystemProvider.ts:225]:
```suggestion
      const serverName = isCSPFile(uri) ? uri.path : uri.path.slice(1).replace(/\//g, ".");
```
This removes an unneeded dependency

File: src/providers/FileSystemProvider/FileSystemProvider.ts [hunk 4]
```diff
@@ -184,13 +192,55 @@ export class FileSystemProvider implements vscode.FileSystemProvider {
...
+      if (serverName.slice(-4).toLowerCase() == ".cls") {
+        if (await isClassDeployed(serverName, api)) {
+          result.permissions |= vscode.FilePermission.Readonly;
```
Comment (isc-bsaviano) [inline src/providers/FileSystemProvider/FileSystemProvider.ts:228]: It's a shame the API doesn't have a way to customize the read-only message too

File: src/providers/FileSystemProvider/FileSystemProvider.ts [hunk 5]
```diff
@@ -184,13 +192,55 @@ export class FileSystemProvider implements vscode.FileSystemProvider {
...
+          return result;
+        }
+      }
+
+      // Does server-side source control report it as editable?
+      if (vscode.workspace.getConfiguration("objectscript.serverSourceControl", uri)?.get("respectEditableStatus")) {
+        const query = "select * from %Atelier_v1_Utils.Extension_GetStatus(?)";
+        const statusObj = await api.actionQuery(query, [serverName]);
+        const docStatus = statusObj.result.content.pop();
```
Comment (isc-bsaviano) [inline src/providers/FileSystemProvider/FileSystemProvider.ts:237]:
```suggestion
        const docStatus = statusObj.result?.content?.pop();
```
Can't hurt to be very defensive

File: src/providers/FileSystemProvider/FileSystemProvider.ts (hunk 4 above)
Comment (gjsjohnmurray) [reply src/providers/FileSystemProvider/FileSystemProvider.ts:228]: Indeed

File: package.json (hunk 1 above)
Comment (gjsjohnmurray) [reply package.json:1316]: Done, and also for other boolean properties where it was previously implicit.

**Comment by gjsjohnmurray** [comment]
Set as ready so a testable VSIX gets built.

**Comment by isc-bsaviano** [comment]
Thanks for making those changes

**Comment by isc-bsaviano** [comment]
Should I wait until that linked Pull Request is merged before I test/approve this?

**Comment by gjsjohnmurray** [comment]
The issue the linked PR fixes is only likely to affect how this new feature behaves with Deltanji's server-side source control class, because of how when we cancel a checkout we reset the mtime of the document to that of the reinstated version.

Please go ahead and review it.

**Comment by gjsjohnmurray** [comment]
I pushed a change to remove those defaults.

As for the settings problem, I haven't been able to reproduce this. The setting displays for me in Settings Editor, and when present in JSON it's not marked / tipped as having a problem.

**Review by isc-bsaviano — APPROVED** [review]
@gjsjohnmurray I installed the latest dev vsix and everything looks good. Thanks for this PR!

**Pull request merged**